src/s3/explorer/s3BucketNode.ts

```diff
@@ -18,82 +19,85 @@ import { ErrorNode } from '../../shared/treeview/nodes/errorNode'
+import { inspect } from 'util'
+import { getLogger } from '../../shared/logger'
 
 /**
  * Represents an S3 bucket that may contain folders and/or objects.
  */
 export class S3BucketNode extends AWSTreeNodeBase implements AWSResourceNode, LoadMoreNode {
-    private cache: S3NodeCache
+    private readonly childLoader: ChildNodeLoader
 
-    public constructor(public readonly bucket: Bucket, private readonly s3: S3Client) {
+    public constructor(
+        public readonly bucket: Bucket,
+        private readonly s3: S3Client,
+        private readonly workspace = Workspace.vscode()
+    ) {
         super(bucket.name, vscode.TreeItemCollapsibleState.Collapsed)
         this.tooltip = bucket.name
         this.iconPath = {
-            dark: vscode.Uri.file(ext.iconPaths.dark.s3.bucket),
-            light: vscode.Uri.file(ext.iconPaths.light.s3.bucket),
+            dark: vscode.Uri.file(ext.iconPaths.dark.s3),
+            light: vscode.Uri.file(ext.iconPaths.light.s3),
         }
         this.contextValue = 'awsS3BucketNode'
-        this.cache = new S3NodeCache(new S3MoreResultsNode(this))
+        this.childLoader = new ChildNodeLoader(this, token => this.loadPage(token))
     }
 
     public async getChildren(): Promise<AWSTreeNodeBase[]> {
         return await makeChildrenNodes({
-            getChildNodes: async () => {
-                // Consider locking to avoid possible race conditions
-                if (!this.initialChildrenLoaded()) {
-                    return this.loadInitialChildren()
-                }
-
-                return this.getExistingChildren()
-            },
+            getChildNodes: async () => this.childLoader.getChildren(),
             getErrorNode: async (error: Error) =>
                 new ErrorNode(this, error, localize('AWS.explorerNode.s3.error', 'Error loading S3 resources')),
             getNoChildrenPlaceholderNode: async () =>
-                new PlaceholderNode(this, localize('AWS.explorerNode.s3.noFiles', '[No Files found]')),
+                new PlaceholderNode(this, localize('AWS.explorerNode.s3.noObjects', '[No Objects found]')),
         })
     }
 
-    private initialChildrenLoaded(): boolean {
-        return !this.cache.isPristine
-    }
-
-    private async loadInitialChildren(): Promise<AWSTreeNodeBase[]> {
-        return this.loadMoreChildren()
+    public async loadMoreChildren(): Promise<AWSTreeNodeBase[]> {
+        return this.childLoader.loadMoreChildren()
     }
 
-    private async getExistingChildren(): Promise<AWSTreeNodeBase[]> {
-        return this.cache.nodes
+    public clearChildren(): void {
+        this.childLoader.clearChildren()
     }
 
-    public async loadMoreChildren(): Promise<AWSTreeNodeBase[]> {
+    private async loadPage(continuationToken: string | undefined): Promise<ChildNodePage> {
+        getLogger().debug(`Loading page for %O using continuationToken %s`, this, continuationToken)
         const response = await this.s3.listObjects({
```

Comment:
Just making sure I'm clear, what's the error handling strategy?

Reply:
Good callout. The initial page failure inserts an error node. But if a subsequent load more fails, it looks like the button doesn't do anything. 

`loadMoreChildren` failures probably need to trigger an error message to let the user know. Doesn't seem appropriate to put the error node in there.